src/protocol/sort/shuffle.rs

```diff
@@ -0,0 +1,280 @@
+        );
+
+        let sequence2left = Shuffle::<Fp31, TestStep>::generate_random_sequence(
+            batchsize,
+            random_value,
+            true,
+            &p2[TestStep::Shuffle(Step1)],
+        );
+        let mut sequence2right = Shuffle::<Fp31, TestStep>::generate_random_sequence(
+            batchsize,
+            random_value,
+            false,
+            &p2[TestStep::Shuffle(Step1)],
+        );
+
+        let sequence3left = Shuffle::<Fp31, TestStep>::generate_random_sequence(
+            batchsize,
+            random_value,
+            true,
+            &p3[TestStep::Shuffle(Step1)],
+        );
+        let mut sequence3right = Shuffle::<Fp31, TestStep>::generate_random_sequence(
+            batchsize,
+            random_value,
+            false,
+            &p3[TestStep::Shuffle(Step1)],
+        );
+
+        let expected_numbers: Vec<usize> = (0..batchsize).collect();
+        assert_eq!(sequence1right, sequence2left);
+        assert_eq!(sequence2right, sequence3left);
+        assert_eq!(sequence3right, sequence1left);
+
+        sequence1right.sort_unstable();
+        sequence2right.sort_unstable();
+        sequence3right.sort_unstable();
+
+        assert_eq!(sequence1right, expected_numbers);
+        assert_eq!(sequence2right, expected_numbers);
+        assert_eq!(sequence3right, expected_numbers);
+    }
+
+    #[tokio::test]
+    async fn shuffle() {
+        let world: TestWorld<TestStep> = make_world(QueryId);
+        let context = make_contexts(&world);
+        let mut rand = StepRng::new(100, 1);
+
+        let input: HashSet<u8> = HashSet::from([4, 1, 9, 2, 3, 8, 6, 7, 5]);
+        let input_len = input.len();
+
+        let mut shares0 = Vec::with_capacity(input_len);
+        let mut shares1 = Vec::with_capacity(input_len);
+        let mut shares2 = Vec::with_capacity(input_len);
+
+        input.clone().into_iter().for_each(|iter| {
+            let share = share(Fp31::from(iter), &mut rand);
+            shares0.push(share[0]);
+            shares1.push(share[1]);
+            shares2.push(share[2]);
+        });
+
+        let mut sh0 = Shuffle::new(shares0, TestStep::Shuffle);
+        let mut sh1 = Shuffle::new(shares1, TestStep::Shuffle);
+        let mut sh2 = Shuffle::new(shares2, TestStep::Shuffle);
+
+        let shuffle0 = sh0.execute(&context[0]);
+        let shuffle1 = sh1.execute(&context[1]);
+        let shuffle2 = sh2.execute(&context[2]);
+
+        let result = try_join!(shuffle0, shuffle1, shuffle2).unwrap();
+
+        assert_eq!(result.0.len(), input_len);
+        assert_eq!(result.1.len(), input_len);
+        assert_eq!(result.2.len(), input_len);
+
+        let mut output = HashSet::new();
+        (0..result.0.len()).for_each(|i| {
+            let val = validate_and_reconstruct((result.0[i], result.1[i], result.2[i]));
+            output.insert(u8::from(val));
+        });
+
+        assert_eq!(output, input);
```

Comment:
This only asserts that the same elements exist in the shuffled list. We need additional checks to validate that the order of the elements is different, and each element has been re-shared in a way that is different from the original.